generators/php/codegen/src/ast/Comment.ts

```diff
@@ -0,0 +1,70 @@
+import { AstNode } from "./core/AstNode";
+import { Writer } from "./core/Writer";
+import { Type } from "./Type";
+
+export type TagType = "param" | "returns" | "throws" | "var";
+
+export const TagType = {
+    Param: "param",
+    Returns: "returns",
+    Throws: "throws",
+    Var: "var"
+} as const;
+
+export declare namespace Comment {
+    interface Args {
+        /* The preface docs of the comment, if any */
+        docs?: string;
+    }
+
+    interface Tag {
+        /* The type of the comment tag (e.g. @param) */
+        tagType: TagType;
+        /* The type included in the @<tag> comment */
+        type: Type;
+        /* The name of the variable in the @<tag> comment, if any */
+        name?: string;
+        /* The in-line docs associated with the type, if any */
+        docs?: string;
+    }
+}
+
+export class Comment extends AstNode {
+    public readonly docs: string | undefined;
+
+    private tags: Comment.Tag[] = [];
+
+    constructor({ docs }: Comment.Args = {}) {
+        super();
+        this.docs = docs;
+    }
+
+    public addTag(tag: Comment.Tag): void {
+        this.tags.push(tag);
+    }
+
+    public write(writer: Writer): void {
+        writer.writeLine("/**");
+        if (this.docs != null) {
+            this.docs.split("\n").forEach((line) => {
+                writer.writeLine(`* ${line}`);
+            });
+        }
+        for (const tag of this.tags) {
+            this.writeTag({ writer, tag });
+        }
+        writer.writeLine(" */");
+    }
+
+    private writeTag({ writer, tag }: { writer: Writer; tag: Comment.Tag }): void {
+        writer.write(` * @${tag.tagType} `);
+        tag.type.write(writer, undefined, true); // Write the tag type as a comment.
```

Comment:
this would be easier to read if we used dictionaries `tag.type.write(writer, { comment: true })`

Reply:
Nice! I actually thought about this for a bit, and ended up sticking with it as is because I didn't want to sacrifice the positional writer parameter. Splitting out the optional into a separate dictionary is exactly what we should do 👍

Reply:
The signature would be something like 

```
public void write(writer: Writer, opts: { comment: boolean } = { comment: false }); 
```